apm-agent-core/src/main/java/co/elastic/apm/agent/util/PackageScanner.java

```diff
@@ -48,6 +49,34 @@ public class PackageScanner {
      * @throws URISyntaxException
      */
     public static List<String> getClassNames(final String basePackage, ClassLoader classLoader) throws IOException, URISyntaxException {
+        // This method is called from indy instrumentation call site method resolution, which happens within the application
+        // threads. Those threads being owned by the application, they could have been interrupted and have their
+        // 'interrupted' status still set, which in turn makes the IO read throw 'ClosedByInterruptException'.
+        //
+        // When this happens, we can clear the 'interrupted' status, retry once and then restore the 'interrupted'
+        // state as it was before this method was called.
+
+        List<String> list;
+        try {
+            list = doGetClassNames(basePackage, classLoader);
+        } catch (ClosedByInterruptException e) {
+            // clears the 'interrupted' status, expected to return true as exception was thrown
+            if (Thread.interrupted()) {
+
+                list = doGetClassNames(basePackage, classLoader);
+
+                // restore the 'interrupted' status
+                Thread.currentThread().interrupt();
+            } else {
+                throw new IllegalStateException("unexpected thread state", e);
+            }
+
```

Comment:
If the interrupt status has been cleared in the meantime, we don't need to throw a `IllegalStateException`, I think.

```suggestion
            boolean interrupted = Thread.interrupted();
            try {
                list = doGetClassNames(basePackage, classLoader);
            } finally {
                if (interrupted) {
                  // restore the 'interrupted' status
                  Thread.currentThread().interrupt();
                }
            }
```